Allow multiple modifier characters in io:format control sequences

richcarl opened the pull request:
Currently, the t and l modifiers are not allowed at the same time. This means that unicode atoms cannot be printed at the same time as suppressing detection of printable lists:
``` 
    13> io:fwrite("~lp~n", [{[666],list_to_atom([666])}]).
    {[666],'\x{29A}'}
    14> io:fwrite("~tp~n", [{[666],list_to_atom([666])}]).
    {"ʚ",'ʚ'}
```
with this patch, you can also do:
```
    15> io:fwrite("~ltp~n", [{[666],list_to_atom([666])}]).
    {[666],'ʚ'}
```
Note: the erl_lint:extract_sequence() function could probably be completely rewritten in a simpler and nicer way using the io_lib:scan_format() functionality, but I don't have time for such things right now.

uabboli commented on lib/stdlib/doc/src/io.xml:
```diff
@@ -167,13 +167,26 @@ ok</pre>
+        <p>To use a literal <c>*</c> character as <c>Pad</c>, it must be
+          passed as an argument:</p>
+        <pre>
+2> <input>io:fwrite("~*.*.0f~n",[9, 5, $*, 3.14159265]).</input>
```
"*.*.*f"

uabboli replied:
Ahh, I meant that the zero should be a star.

richcarl replied:
Bah, serves me right for not copy/pasting the entire line from the shell after testing.

uabboli commented on lib/stdlib/src/erl_lint.erl:
```diff
@@ -3984,13 +3986,17 @@ extract_sequence(4, [$t, $w | Fmt], Need) ->
+    extract_sequence(5, [$P|Fmt], Need);
 extract_sequence(4, [$l, C | _Fmt], _Need) ->
-    {error,"invalid control ~l" ++ [C]};
+    {error,"invalid modifier ~l" ++ [C]};
```
C is no longer always a control. ltc in a format string, for example.

richcarl replied:
Right, I missed that the error message appends the rest of the string. Restoring it to the old way (but if this code gets rewritten, it would be nice to point out more exactly that the problem is that the modifier is not allowed with the main control character).

uabboli commented:
Everything looks good except for one testcase, io_SUITE:otp_10755/1, line 1913 and forward.

richcarl commented:
Rewrote the test case to cover all the new allowed variants and changed the check for compiler warning about ~lt combination to warn about using ~l or ~lt or ~tl with ~w/~W.

richcarl commented:
Done. I had followed the example that matched on the errors, but that's obviously not needed now that no tests actually raise exceptions.
